lab3van/util-linux/sys-utils/fsfreeze.8
<<<<<<< HEAD
.\" -*- nroff -*-
.TH FSFREEZE 8 "May 2010" "util-linux" "System Administration"
.SH NAME
fsfreeze \- suspend access to a filesystem (Linux Ext3/4, ReiserFS, JFS, XFS)
.SH SYNOPSIS
.B fsfreeze \-f
.I mountpoint

.B fsfreeze \-u
.I mountpoint

.SH DESCRIPTION
.B fsfreeze
suspends and resumes access to an filesystem
.PP
.B fsfreeze
halts new access to the filesystem and creates a stable image on disk.
.B fsfreeze
is intended to be used with hardware RAID devices that support the creation 
=======
.TH FSFREEZE 8 "July 2014" "util-linux" "System Administration"
.SH NAME
fsfreeze \- suspend access to a filesystem (Ext3/4, ReiserFS, JFS, XFS)
.SH SYNOPSIS
.B fsfreeze
.BR \--freeze | \--unfreeze
.IR mountpoint

.SH DESCRIPTION
.B fsfreeze
suspends or resumes access to a filesystem.
.PP
.B fsfreeze
halts any new access to the filesystem and creates a stable image on disk.
.B fsfreeze
is intended to be used with hardware RAID devices that support the creation
>>>>>>> master-vanilla
of snapshots.
.PP
.B fsfreeze
is unnecessary for
.B device-mapper
<<<<<<< HEAD
devices. The device-mapper (and LVM)
automatically freezes filesystem on the device when a snapshot creation is requested.
=======
devices.  The device-mapper (and LVM) automatically freezes a filesystem
on the device when a snapshot creation is requested.
>>>>>>> master-vanilla
For more details see the
.BR dmsetup (8)
man page.
.PP
The
<<<<<<< HEAD
.I mount-point
=======
.I mountpoint
>>>>>>> master-vanilla
argument is the pathname of the directory where the filesystem
is mounted.
The filesystem must be mounted to be frozen (see
.BR mount (8)).
<<<<<<< HEAD
.SH OPTIONS
.IP "\fB\-h, \-\-help\fP"
Print help and exit.
.IP "\fB\-f, \-\-freeze\fP"
=======
.PP
Note that access-time updates are also suspended if the filesystem is mounted with
the traditional atime behavior (mount option \fBstrictatime\fR, for more details see
.BR mount (8)).

.SH OPTIONS
.TP
.BR \-f , " \-\-freeze"
>>>>>>> master-vanilla
This option requests the specified a filesystem to be frozen from new
modifications.  When this is selected, all ongoing transactions in the
filesystem are allowed to complete, new write system calls are halted, other
calls which modify the filesystem are halted, and all dirty data, metadata, and
log information are written to disk.  Any process attempting to write to the
frozen filesystem will block waiting for the filesystem to be unfrozen.
<<<<<<< HEAD

=======
.sp
>>>>>>> master-vanilla
Note that even after freezing, the on-disk filesystem can contain
information on files that are still in the process of unlinking.
These files will not be unlinked until the filesystem is unfrozen
or a clean mount of the snapshot is complete.
<<<<<<< HEAD
.IP "\fB\-u, \-\-unfreeze\fP
This option is used to un-freeze the filesystem and allow operations to
continue.  Any filesystem modifications that were blocked by the freeze are
unblocked and allowed to complete.
=======
.TP
.BR \-u , " \-\-unfreeze"
This option is used to un-freeze the filesystem and allow operations to
continue.  Any filesystem modifications that were blocked by the freeze are
unblocked and allowed to complete.
.TP
.BR \-V , " \-\-version"
Display version information and exit.
.TP
.BR \-h , " \-\-help"
Display help text and exit.
>>>>>>> master-vanilla
.SH AUTHOR
.PP
Written by Hajime Taira.
.SH NOTES
.PP
<<<<<<< HEAD
This man page based on xfs_freeze.
One of
.B \-f
or
.B \-u
must be supplied to
.BR fsfreeze .
=======
This man page is based on
.BR xfs_freeze (8).
>>>>>>> master-vanilla
.SH SEE ALSO
.BR mount (8)
.SH AVAILABILITY
The fsfreeze command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.
<<<<<<< HEAD

=======
>>>>>>> master-vanilla


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/sys-utils/fsfreeze.c
/*
 * fsfreeze.c -- Filesystem freeze/unfreeze IO for Linux
 *
 * Copyright (C) 2010 Hajime Taira <htaira@redhat.com>
 *                    Masatake Yamato <yamato@redhat.com>
 *
 * This program is free software.  You can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation: either version 1 or
 * (at your option) any later version.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <linux/fs.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <getopt.h>

<<<<<<< HEAD
#include "blkdev.h"
#include "nls.h"
#include "c.h"
=======
#include "c.h"
#include "blkdev.h"
#include "nls.h"
#include "closestream.h"
#include "optutils.h"

enum fs_operation {
	NOOP,
	FREEZE,
	UNFREEZE
};
>>>>>>> master-vanilla

static int freeze_f(int fd)
{
	return ioctl(fd, FIFREEZE, 0);
}

static int unfreeze_f(int fd)
{
	return ioctl(fd, FITHAW, 0);
}

static void __attribute__((__noreturn__)) usage(FILE *out)
{
<<<<<<< HEAD
	fputs(_("\nUsage:\n"), out);
	fprintf(out,
	      _(" %s [options] <mount point>\n"), program_invocation_short_name);

	fputs(_("\nOptions:\n"), out);
	fputs(_(" -h, --help        this help\n"
		" -f, --freeze      freeze the filesystem\n"
		" -u, --unfreeze    unfreeze the filesystem\n"), out);

	fputs(_("\nFor more information see fsfreeze(8).\n"), out);
=======
	fprintf(out, USAGE_HEADER);
	fprintf(out,
	      _(" %s [options] <mountpoint>\n"), program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Suspend access to a filesystem (ext3/4, ReiserFS, JFS, XFS).\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -f, --freeze      freeze the filesystem\n"), out);
	fputs(_(" -u, --unfreeze    unfreeze the filesystem\n"), out);
	fprintf(out, USAGE_SEPARATOR);
	fprintf(out, USAGE_HELP);
	fprintf(out, USAGE_VERSION);
	fprintf(out, USAGE_MAN_TAIL("fsfreeze(8)"));
>>>>>>> master-vanilla

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	int fd = -1, c;
<<<<<<< HEAD
	int freeze = -1, rc = EXIT_FAILURE;
=======
	int action = NOOP, rc = EXIT_FAILURE;
>>>>>>> master-vanilla
	char *path;
	struct stat sb;

	static const struct option longopts[] = {
	    { "help",      0, 0, 'h' },
	    { "freeze",    0, 0, 'f' },
	    { "unfreeze",  0, 0, 'u' },
<<<<<<< HEAD
	    { NULL,        0, 0, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	while ((c = getopt_long(argc, argv, "hfu", longopts, NULL)) != -1) {
=======
	    { "version",   0, 0, 'V' },
	    { NULL,        0, 0, 0 }
	};

	static const ul_excl_t excl[] = {       /* rows and cols in in ASCII order */
		{ 'f','u' },			/* freeze, unfreeze */
		{ 0 }
	};
	int excl_st[ARRAY_SIZE(excl)] = UL_EXCL_STATUS_INIT;

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	while ((c = getopt_long(argc, argv, "hfuV", longopts, NULL)) != -1) {

		err_exclusive_options(c, longopts, excl, excl_st);

>>>>>>> master-vanilla
		switch(c) {
		case 'h':
			usage(stdout);
			break;
		case 'f':
<<<<<<< HEAD
			freeze = TRUE;
			break;
		case 'u':
			freeze = FALSE;
			break;
=======
			action = FREEZE;
			break;
		case 'u':
			action = UNFREEZE;
			break;
		case 'V':
			printf(UTIL_LINUX_VERSION);
			exit(EXIT_SUCCESS);
>>>>>>> master-vanilla
		default:
			usage(stderr);
			break;
		}
	}

<<<<<<< HEAD
	if (freeze == -1)
		errx(EXIT_FAILURE, _("no action specified"));
=======
	if (action == NOOP)
		errx(EXIT_FAILURE, _("neither --freeze or --unfreeze specified"));
>>>>>>> master-vanilla
	if (optind == argc)
		errx(EXIT_FAILURE, _("no filename specified"));
	path = argv[optind++];

	if (optind != argc) {
		warnx(_("unexpected number of arguments"));
		usage(stderr);
	}

	fd = open(path, O_RDONLY);
	if (fd < 0)
<<<<<<< HEAD
		err(EXIT_FAILURE, _("%s: open failed"), path);

	if (fstat(fd, &sb) == -1) {
		warn(_("%s: fstat failed"), path);
=======
		err(EXIT_FAILURE, _("cannot open %s"), path);

	if (fstat(fd, &sb) == -1) {
		warn(_("stat of %s failed"), path);
>>>>>>> master-vanilla
		goto done;
	}

	if (!S_ISDIR(sb.st_mode)) {
		warnx(_("%s: is not a directory"), path);
		goto done;
	}

<<<<<<< HEAD
	if (freeze) {
=======
	switch (action) {
	case FREEZE:
>>>>>>> master-vanilla
		if (freeze_f(fd)) {
			warn(_("%s: freeze failed"), path);
			goto done;
		}
<<<<<<< HEAD
	} else {
=======
		break;
	case UNFREEZE:
>>>>>>> master-vanilla
		if (unfreeze_f(fd)) {
			warn(_("%s: unfreeze failed"), path);
			goto done;
		}
<<<<<<< HEAD
=======
		break;
	default:
		abort();
>>>>>>> master-vanilla
	}

	rc = EXIT_SUCCESS;
done:
	if (fd >= 0)
		close(fd);
	return rc;
}

